server/src/main/java/org/apache/iotdb/db/metadata/logfile/MLogWriter.java

```diff
@@ -0,0 +1,424 @@
+      // if both .bin and .bin.tmp do not exist, nothing to do
+    } else if (!logFile.exists() && tmpLogFile.exists()) {
+      // if old .bin doesn't exist but .bin.tmp exists, rename tmp file to .bin
+      FSFactoryProducer.getFSFactory().moveFile(tmpLogFile, logFile);
+    } else if (tmpLogFile.exists()) {
+      // if both .bin and .bin.tmp exist, delete .bin.tmp
+      try {
+        Files.delete(Paths.get(tmpLogFile.toURI()));
+      } catch (IOException e) {
+        throw new IOException("Deleting " + tmpLogFile + "failed with exception " + e.getMessage());
+      }
+    }
+
+    // do some clean job
+    // remove old .txt and .txt.tmp
+    if (oldLogFile.exists()) {
+      try {
+        Files.delete(Paths.get(oldLogFile.toURI()));
+      } catch (IOException e) {
+        throw new IOException("Deleting " + oldLogFile + "failed with exception " + e.getMessage());
+      }
+    }
+
+    if (tmpOldLogFile.exists()) {
+      try {
+        Files.delete(Paths.get(tmpOldLogFile.toURI()));
+      } catch (IOException e) {
+        throw new IOException("Deleting " + tmpOldLogFile + "failed with exception " + e.getMessage());
+      }
+    }
+
+    // rename .bin.tmp to .bin
+    FSFactoryProducer.getFSFactory().moveFile(tmpLogFile, logFile);
+  }
+
+  public static void upgradeMLog() throws IOException {
+    String schemaDir = IoTDBDescriptor.getInstance().getConfig().getSchemaDir();
+    upgradeTxtToBin(schemaDir, MetadataConstant.METADATA_TXT_LOG, MetadataConstant.METADATA_LOG, false);
+    upgradeTxtToBin(schemaDir, MetadataConstant.MTREE_TXT_SNAPSHOT, MetadataConstant.MTREE_SNAPSHOT, true);
+  }
+
+  public void clear() throws IOException {
+    sync();
+    logWriter.close();
+    mlogBuffer.clear();
+    if (logFile != null && logFile.exists()) {
+      Files.delete(logFile.toPath());
+    }
+    logNum = 0;
+    logWriter = new LogWriter(logFile, 0L);
+  }
+
+  public int getLogNum() {
+    return logNum;
+  }
+
+  /**
+   * only used for initialize a mlog file writer.
+   */
+  public void setLogNum(int number) {
+    logNum = number;
+  }
+
+  public void operation(String cmd, boolean isSnapshot) throws IOException, MetadataException {
+    if (!isSnapshot) {
+      operation(cmd);
+    } else {
+      PhysicalPlan plan = convertFromString(cmd);
+      try {
+        if (plan != null) {
+          putLog(plan);
+        }
+      } catch (BufferOverflowException e) {
+        throw new IOException(
+          LOG_TOO_LARGE_INFO, e);
+      }
+    }
+  }
+
+  /**
+   * upgrade from mlog.txt to mlog.bin
+   * @param cmd, the old meta operation
+   * @throws IOException
+   * @throws MetadataException
+   */
+  public void operation(String cmd) throws IOException, MetadataException {
+    // see createTimeseries() to get the detailed format of the cmd
+    String[] args = cmd.trim().split(",", -1);
+    switch (args[0]) {
+      case MetadataOperationType.CREATE_TIMESERIES:
```

Comment:
If there is a comma in the path of a timeseries, the length of args[] will be larger than 8.
I think adding these lines back is better.
```
        if (args.length > 8) {	
          String[] tmpArgs = new String[8];	
          tmpArgs[0] = args[0];	
          int i = 1;	
          tmpArgs[1] = "";	
          for (; i < args.length - 7; i++) {	
            tmpArgs[1] += args[i] + ",";	
          }	
          tmpArgs[1] += args[i++];	
          for (int j = 2; j < 8; j++) {	
            tmpArgs[j] = args[i++];	
          }	
          args = tmpArgs;	
        }
```

Reply:
ok